rfcs/00xx_UMA_Unified_Modular_Accelerator_Interface.md

```diff
@@ -0,0 +1,184 @@
+  * Output: Target .c files
+
+The intention is to use TensorIR and Relax with MetaScheduler for optimization.
+
+
+Abbreviations:
+S-TIR: Schedulable TIR
+NS-TIR: Non-Schedulable TIR
+
+
+### File and class structure and Snippets as example for integration
+UMA provides a mostly python-based API. On the C++ side, new targets are registered using target hooks (RFC #0010). A generic `codegen.cc` handles the calls to the python side.
+```
+.
+├── codegen.cc
+└── targets.cc
+```
+```cpp
+TVM_REGISTER_TARGET_KIND("accelerator_A", kDLCPU)
+    .set_attr<FTVMRelayToTIR>("RelayToTIR", relay::contrib::generic::RelayToTIR("accelerator_A"))
+    .set_attr<FTVMTIRToRuntime>("TIRToRuntime", relay::contrib::generic::accelerator_A::TIRToRuntime);
+
+TVM_REGISTER_TARGET_KIND("accelerator_B", kDLCPU)
+    .set_attr<FTVMRelayToTIR>("RelayToTIR", relay::contrib::generic::RelayToTIR("accelerator_B"));
+    .set_attr<FTVMTIRToRuntime>("TIRToRuntime", relay::contrib::generic::accelerator_B::TIRToRuntime);
+```
+The python API is structured as shown below. The base class `UMABackend` functions as the core API. It uses the API points described in the [flow description](#flow-description) to register elements for the different stages.
+```
+.
+├── backend.py
+├── api
+│   ├── codegen.py
+│   ├── lower.py
+│   ├── partitioner.py
+│   └── utils.py
+├── accelerator_A
+│   ├── backend.py
+│   ├── codegen.py
+│   ├── passes.py
+│   ├── patterns.py
+│   ├── schedules.py
+│   └── strategies.py
+└── accelerator_B
+    └── ...
+```
+A custom accelerator is added by inheriting the `UMABackend`. New elements (e.g., passes, schedules) are added using a registration machanism.
+```python
+"""UMA backend for the UltraTrail accelerator"""
+
+class UltraTrailBackend(UMABackend):
+    def __init__(self):
+        super(UltraTrailBackend, self).__init__()
+
+        # Relay to Relay function registration
+        self._register_pattern("conv1d_relu", conv1d_relu_pattern())
+
+        self._register_relay_pass(1, ConfigGenerator())
+        self._register_relay_pass(2, BufferScopeAnnotator())
+
+        # Relay to TIR function registration
+        self._register_operator_strategy("nn.conv1d", custom_conv1d_strategy, plevel=9)
+
+        self._register_tir_schedule(insert_extern_calls)
+
+        self._register_tir_pass(0, CodegenGenerateConfig())
+        self._register_tir_pass(0, CodegenGenerateConstants())
+
+        # TIR to runtime function registration
+        self._register_codegen(format="c", includes=None, constants=None, replace_call_extern=None)
+
+    @property
+    def target_name(self):
+        return "ultra_trail"
+```
+Once a `UMABackend` is registered, it hooks into the usual `relay.build` process to create the code for the target accelerator.
+```
+# Load model
+mod, params = relay.frontend.from_pytorch(scripted_model, [("input_data", input_shape)])
+
+# Register a UMA backend
+UltraTrailBackend().register()
```

Comment:
Is this the only thing the user need to do to register the backend ? 
Will it do something to the effect : 

```
TVM_REGISTER_TARGET_KIND("accelerator_B", kDLCPU)
    .set_attr<FTVMRelayToTIR>("RelayToTIR", relay::contrib::generic::RelayToTIR("accelerator_B"));
    .set_attr<FTVMTIRToRuntime>("TIRToRuntime", relay::contrib::generic::accelerator_B::TIRToRuntime);
```

Reply:
> Is this the only thing the user need to do to register the backend ? Will it do something to the effect :

Yes. 
 
> ```
> TVM_REGISTER_TARGET_KIND("accelerator_B", kDLCPU)
>     .set_attr<FTVMRelayToTIR>("RelayToTIR", relay::contrib::generic::RelayToTIR("accelerator_B"));
>     .set_attr<FTVMTIRToRuntime>("TIRToRuntime", relay::contrib::generic::accelerator_B::TIRToRuntime);
> ```


`backend.register` interacts with TVM core in 2 ways.
1. It registers the target hooks as you said. 
2. It registers the target pattern tables using `tvm.relay.op.contrib.register.register_pattern_table` and ensures that both use the same compiler/target_name .

Reply:
Hmmmmm.. TVM_REGISTER_TARGET_KIND is a C/C++ macro. Do we envision to have some sort of python decorator (as part of this work) to handle the registration?

Reply:
@manupa-arm: This part was removed. The C++ code you mention is no longer needed for target registration.

CC: @cgerum

Reply:
Just a short addition regarding the implementation. We are using a global function `RegisterTarget` in C++, that takes the target name and registers the target together with the target hooks. `RegisterTarget` is called during the backend registration `UltraTrailBackend().register()`. To hide this process from the user we are not using a decorator, but I think it's a similar approach.

Reply:
That make sense to me! It might be worth adding this to the Reference explanation :) 

How/Do we deal with target specific options ?

E.g. : 
https://github.com/apache/tvm/blob/fe7b5d329a82f720a721356c40abd721cf1d780d/src/target/target_kind.cc#L373

Reply:
I added it to the Reference explanation [(link)](https://github.com/boschresearch/tvm-rfcs/blob/rfc_uma/rfcs/00xx_UMA_Unified_Modular_Accelerator_Interface.md#uma-target-hooks).

Currently it is not possible to set target specific options. Since the UMA targets are essentially also "c" targets we did not see the need to deal with target specific options. Do you have a use-case in mind, for which this would be necessary?

Reply:
Well, we currently use such options to define accelerator variants that share the same lowering pipeline.
In the absense of that, we would need to resort to use the PassConfig, however, generally PassConfig is generally better suited to set a configuration to a specific Pass. In the above case, it would mean we need to set multiple Passes in the absense.

I would see the newly registered targets as extensions of the "c" target and Im a bit keen on not ending up having to dump a union of UMA target options to "c" target. 

Following your proposal, is there a reason why we wont be able to use RegisterTarget ? We could consider including AttrDict to that effect.

Adding to this, there are two variants of these : 
 
 * relay.ext.<backend>.options
 
 Which define the options for the lowering. This is inherited by the original BYOC design and we still use it with Target Hooks. This is partly due to the seperate existence kCompiler strings and actual targets.
 
 * target_kind options
 
 This is what I alluded to in the previous comment.
 
Ideally, since UMA is wrapping Target Hooks, I suppose if we want to add this, we would want to proceed with the second option -- hence the suggestion.

Reply:
> Adding to this, there are two variants of these :
> 
> * relay.ext.<backend>.options
> 
> Which define the options for the lowering. This is inherited by the original BYOC design and we still use it with Target Hooks. This is partly due to the seperate existence kCompiler strings and actual targets.

Yes this not exactly user friendly.

> * target_kind options
> 
> This is what I alluded to in the previous comment.
> 
> Ideally, since UMA is wrapping Target Hooks, I suppose if we want to add this, we would want to proceed with the second option -- hence the suggestion.

Adding them to the TargetKind would obviously be the preferred solution. As would potentially be a preferred way of doing things. It would allow for a clean `target = Target("ethos-u -version=ethosu-256", host=Target("c"))`  but this makes the current partitioning Flow somewhat clumsy as we still need to run `.partition` for the target hooks. On the other hand it could serve as a starting point for deprecating `partition_for` and providing a unified API for Collage and non-collage flows. @mbs-octoml @areusch would that make sense?